msal/application.py

```diff
@@ -1046,6 +1050,21 @@ def _forget_me(self, home_account):
                 TokenCache.CredentialType.ACCOUNT, query=owned_by_home_account):
             self.token_cache.remove_account(a)
 
+    def _acquire_token_by_cloud_shell(self, scopes, data=None):
+        from .cloudshell import _acquire_token
+        response = _acquire_token(
+            self.http_client, scopes, client_id=self.client_id, data=data)
+        if "error" not in response:
+            self.token_cache.add(dict(
+                client_id=self.client_id,
+                scope=response["scope"].split() if "scope" in response else scopes,
+                token_endpoint=self.authority.token_endpoint,
+                response=response.copy(),
+                data=data or {},
+                authority_type=_AUTHORITY_TYPE_CLOUDSHELL,
+                ))
```

Comment:
I think Cloud Shell's pseudo managed identity endpoint already has a cache, so does a normal managed identity endpoint. Do we really need to save it to MSAL's cache?

Reply:
The devil is in the details.

> I think Cloud Shell's pseudo managed identity endpoint already has a cache,

The SSH Cert feature has its own quirk in caching. Currently, the Cloud Shell (actually the Azure Portal) does not yet really support caching for SSH Cert. 

> so does a normal managed identity endpoint.

Yes but there is a catch. Quoted from [normal managed identity's document](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#token-caching): "The managed identities subsystem caches tokens but we still recommend that you implement token caching in your code." They even defined [an HTTP 429 throttling error when you use Managed Identity endpoint too often](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#error-handling). FWIW, I'm currently working on another project that is trying to enable MSAL token cache for managed identity.

> Do we really need to save it to MSAL's cache?

The MSAL cache has long been designed to be used as often as you want, and it also stores SSH cert. Adding this (technically still a) one-liner here can allow existing MSAL-powered apps' `acquire_token_silent()` to work, for free (meaning without any extra code change in app's code). Why not?